File: 234-palindrome-linked-list/palindrome-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public boolean isPalindrome(ListNode head) {
     if(head == null || head.next == null) return true;

     ListNode fast = head, slow = head;

     while(fast!=null && fast.next!=null){
        fast = fast.next.next;
        slow = slow.next;
     }

     ListNode firstHalf = head, secondHalf = reverse(slow);

     while(secondHalf != null){
        if(firstHalf.val != secondHalf.val){
            return false;
        }
        firstHalf = firstHalf.next;
        secondHalf= secondHalf.next;
     }
     return true;
    }

    ListNode reverse(ListNode head){
        if(head == null || head.next == null) return head;

        ListNode reverseHead = reverse(head.next);

        head.next.next = head;
        head.next = null;

        return reverseHead;
    }
}

File: 234-palindrome-linked-list/palindrome-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        # arr = []

        # temp = head
        # while temp:
        #     arr.append(temp.val)
        #     temp = temp.next

        # return arr == arr[::-1]

        slow, fast = head, head

        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next

        prev = None
        curr = slow

        while curr:
            nxt = curr.next
            curr.next = prev
            prev = curr
            curr = nxt

        left, right = head, prev

        while right:
            if left.val != right.val:
                return False

            left = left.next
            right = right.next

        return True


        
        